보내주신 pull request를 ubuntu 12.04에서 설치해보았습니다.
도커를 사용하지 않고 새로운 ubuntu 12.04 버전의 가상머신을 생성하여 테스트 하였으며
arcus를 설치하고 빌드까지 마친 후에 올려주신 pull request를 모두 적용, forever에는 sudo 권한을 주어 설치하였습니다.(올려주신 package 버전 수정 request도 적용했습니다.)
하지만 여전히 에러가 발생합니다.
혹시 몰라 기존 권고한 node버전 외에 추가로 0.10.45, 0.12.7 버전에서도 실행해보았습니다.

```
npm ERR! addLocal Could not install /home/hubble/arcus-collectd/lib/node_rrd
npm ERR! Error: ENOENT, stat '/home/hubble/arcus-collectd/lib/node_rrd'
npm ERR! If you need help, you may report this *entire* log,
npm ERR! including the npm and node versions, at:
npm ERR!     <http://github.com/npm/npm/issues>

npm ERR! System Linux 3.13.0-32-generic
npm ERR! command "/home/hubble/vendor/node/bin/node" "/home/hubble/vendor/node/bin/npm" "install" "/home/hubble/arcus-collectd/lib/node_rrd"
npm ERR! cwd /home/hubble/arcus-hubble
npm ERR! node -v v0.10.28
npm ERR! npm -v 1.4.9
npm ERR! path /home/hubble/arcus-collectd/lib/node_rrd
npm ERR! code ENOENT
npm ERR! errno 34
npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     /home/hubble/arcus-hubble/npm-debug.log
npm ERR! not ok code 0
```

```
npm ERR! addLocal Could not install /home/hubble/arcus-collectd/lib/node_rrd
npm ERR! Linux 3.13.0-32-generic
npm ERR! argv "/home/hubble/vendor/node/bin/node" "/home/hubble/vendor/node/bin/npm" "install" "/home/hubble/arcus-collectd/lib/node_rrd"
npm ERR! node v0.10.45
npm ERR! npm  v2.15.1
npm ERR! path /home/hubble/arcus-collectd/lib/node_rrd
npm ERR! code ENOENT
npm ERR! errno 34

npm ERR! enoent ENOENT, open '/home/hubble/arcus-collectd/lib/node_rrd'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /home/hubble/arcus-hubble/npm-debug.log
```

<br>

또한 centos7에서도 다음과 같은 에러가 발생했습니다.

```
npm ERR! addLocal Could not install /home/hubble/arcus-collectd/lib/node_rrd
npm ERR! Linux 3.10.0-327.18.2.el7.x86_64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "/home/hubble/arcus-collectd/lib/node_rrd"
npm ERR! node v0.10.45
npm ERR! npm  v2.15.1
npm ERR! path /home/hubble/arcus-collectd/lib/node_rrd
npm ERR! code ENOENT
npm ERR! errno 34

npm ERR! enoent ENOENT, open '/home/hubble/arcus-collectd/lib/node_rrd'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /home/hubble/arcus-hubble/npm-debug.log
```

추가적으로 덧붙이자면 깃허브 node_rrd 자체의 문제가 아닐까 싶어 내려받아서 npm install을 진행하자 ubuntu에서는 아래와 같은 에러가 발생했으며 centos에서는 문제가 발생하지 않았습니다.

```
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/hubble/vendor/node-v0.10.45-linux-x64/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System Linux 3.13.0-32-generic
gyp ERR! command "node" "/home/hubble/vendor/node-v0.10.45-linux-x64/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/hubble/node_rrd
gyp ERR! node -v v0.10.45
gyp ERR! node-gyp -v v3.3.1
gyp ERR! not ok

npm ERR! Linux 3.13.0-32-generic
npm ERR! argv "/home/hubble/vendor/node/bin/node" "/home/hubble/vendor/node/bin/npm" "install"
npm ERR! node v0.10.45
npm ERR! npm  v2.15.1
npm ERR! code ELIFECYCLE
npm ERR! rrd@0.0.5 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the rrd@0.0.5 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the rrd package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs rrd
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!
npm ERR!     npm owner ls rrd
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/hubble/node_rrd/npm-debug.log
```

--- a/README.md
+++ b/README.md
@@ -27,24 +27,46 @@ ln -s node-v0.10.28-linux-x64 node
 # set environment variable
 export NODE_HOME=$HOME/vendor/node
 export PATH=$NODE_HOME/bin:$PATH
+
+# Change the permission of npm directories for global packages.
+# https://docs.npmjs.com/getting-started/fixing-npm-permissions
+sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}
+```
+
+- build tools installation
+
+To build native binding libraries (e.g. node_rrd), compile toolchains are required.
+
+```bash
+# Ubuntu
+sudo apt-get install build-essential ca-certificates
+
+# CentOS
+sudo yum groupinstall 'Development Tools'
+```
+
+On the ubuntu 12.04, update gcc and g++. (gcc and g++ should be newer than 4.8 to build node_rrd.)
+
+```bash
+# Ubuntu 12.04
+sudo apt-get install python-software-properties
+sudo add-apt-repository ppa:ubuntu-toolchain-r/test
+sudo apt-get update
+sudo apt-get install gcc-4.8 g++-4.8
+sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.8 50
+sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-4.8 50
 ```
 
 - `rrdtool & collectd` installation
 
 ```
+# Ubuntu
+sudo apt-get install librrd-dev
 
-sudo apt-get install libcairo2-dev libpango1.0-dev libxml2-dev (Ubuntu)
-sudo yum install cairo-devel libcairo2-devel libpango1.0-devel libxml-devel (CentOS)
+# CentOS
+sudo yum install rrdtool-devel
 
 cd ~/vendor
-
-# rrdtool
-curl -OL http://oss.oetiker.ch/rrdtool/pub/rrdtool-1.4.8.tar.gz
-tar xvf rrdtool-1.4.8.tar.gz
-pushd rrdtool-1.4.8
-./configure --prefix=$HOME/arcus-collectd
-make; make install
-popd
 
 # collectd
 curl -OL https://collectd.org/files/collectd-5.4.1.tar.gz
@@ -174,4 +196,3 @@ https://github.com/naver/arcus-hubble/issues
 ## License
 
 Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
-

--- a/lib/hubble-rrd.js
+++ b/lib/hubble-rrd.js
@@ -17,7 +17,7 @@
 var path = require('path');
 var HUBBLE_HOME = path.resolve(__dirname + '/..');
 
-var rrd = require(HUBBLE_HOME + '/lib/node_rrd/lib/rrd.js')
+var rrd = require('rrd')
   , rrddir = require('./rrd/rrddir.js')
   , async = require('async')
   , fs = require('fs')

--- a/lib/node_rrd/binding.gyp
+++ b/lib/node_rrd/binding.gyp
@@ -2,7 +2,7 @@
   "targets": [
     {
       "target_name": "rrd_bindings",
-      "sources": [ 
+      "sources": [
           "src/node_rrd_create.cc"
         , "src/node_rrd_update.cc"
         , "src/node_rrd_fetch.cc"
@@ -10,10 +10,9 @@
       	, "src/node_rrd_info.cc"
       	, "src/node_rrd.cc"
       ],
-      'libraries': [ "-L${COLLECTD_HOME}/lib -lrrd" ],
-      'include_dirs': [
-        '${COLLECTD_HOME}/include'
-      ]
+      'include_dirs' : [ "<!(node -e \"require('nan')\")" ],
+      'libraries': [ "-lrrd" ],
+      'cflags_cc': [ "-std=c++11" ]
     }
   ]
 }

--- a/lib/node_rrd/package.json
+++ b/lib/node_rrd/package.json
@@ -1,9 +1,9 @@
-{ 
+{
     "name": "rrd"
   , "version": "0.0.5"
   , "author": "Thierry Passeron (thierry.passeron@gmail.com)"
   , "contributors": [
-    
+
   ]
   , "description": "Node.js native binding for RRDtool"
   , "keywords": [
@@ -13,14 +13,15 @@
       , "node"
       , "native"
   ]
-  , "engines": { "node": ">=0.6.0" }
+  , "engines": { "node": ">=0.6.0 <0.13" }
   , "main": "./lib/rrd"
   , "scripts": {
       "test": "cd test; make"
   }
   , "dependencies": {
       "bindings": "*"
-    , "arguee": ">=0.0.2" 
+    , "arguee": ">=0.0.2"
+    , "nan": ">=2.0.9"
   }
   , "devDependencies": {
       "mocha": ">=1.3.0"

--- a/lib/node_rrd/src/node_rrd.cc
+++ b/lib/node_rrd/src/node_rrd.cc
@@ -1,6 +1,6 @@
 /*
     RRDtool (http://oss.oetiker.ch/rrdtool/) bindings module for node (http://nodejs.org)
-    
+
     Copyright (c), 2012 Thierry Passeron
 
     The MIT License
@@ -27,13 +27,13 @@
 #include "node_rrd.h"
 
 extern "C" {
-    static void init(Handle<Object> target) {
-        HandleScope scope;
-        NODE_SET_METHOD(target, "create", node_rrd::create);
-        NODE_SET_METHOD(target, "update", node_rrd::update);
-        NODE_SET_METHOD(target, "fetch", node_rrd::fetch);
-        NODE_SET_METHOD(target, "last", node_rrd::last);
-        NODE_SET_METHOD(target, "info", node_rrd::info);
+    static void init(Local<Object> target) {
+        Nan::HandleScope scope;
+        Nan::SetMethod(target, "create", node_rrd::create);
+        Nan::SetMethod(target, "update", node_rrd::update);
+        Nan::SetMethod(target, "fetch", node_rrd::fetch);
+        Nan::SetMethod(target, "last", node_rrd::last);
+        Nan::SetMethod(target, "info", node_rrd::info);
     }
 
     NODE_MODULE(rrd_bindings, init)

--- a/lib/node_rrd/src/node_rrd.h
+++ b/lib/node_rrd/src/node_rrd.h
@@ -36,6 +36,8 @@
 #   include <iostream> 
 #endif
 
+#include <nan.h>
+
 using namespace v8;
 using namespace node;
 
@@ -51,18 +53,18 @@ public:
     int status;
     
     /* Used by the uv after_work callback */
-    Persistent<Function> callback;
+    Nan::Callback *callback;
 
-    ~AsyncInfos() { callback.Dispose(); free(filename); }
+    ~AsyncInfos() { delete callback; free(filename); }
 };
 
 /* Some useful macros to extract arguments */
 #define SET_CHARS_ARG(I, VAR)                                           \
-    String::Utf8Value _chars ## I(args[I]->ToString());                 \
+    String::Utf8Value _chars ## I(info[I]->ToString());                 \
     VAR = strndup(*_chars ## I, _chars ## I.length());
 
 #define SET_ARGC_ARGV_ARG(I, VAR_ARGC, VAR_ARGV)                        \
-    Local<Array> argv ## I = Local<Array>::Cast(args[I]);               \
+    Local<Array> argv ## I = Local<Array>::Cast(info[I]);               \
     VAR_ARGC = argv ## I->Length();                                     \
     VAR_ARGV = (char**)malloc(sizeof(char*) * VAR_ARGC);                \
     for (int i = 0; i < VAR_ARGC; i++) {                                \
@@ -71,17 +73,15 @@ public:
     }
 
 #define CHECK_FUN_ARG(I)                                                \
-  if (args.Length() <= (I) || !args[I]->IsFunction())                   \
-    return v8::ThrowException(v8::Exception::TypeError(                 \
-                  String::New("Argument " #I " must be a function")));  
+  if (info.Length() <= (I) || !info[I]->IsFunction())                   \
+    return Nan::ThrowTypeError("Argument " #I " must be a function");
 
 #define SET_PERSFUN_ARG(I, VAR)                                         \
-  VAR = Persistent<Function>::New(Local<Function>::Cast(args[I]));
+  VAR = new Nan::Callback(Local<Function>::Cast(info[I]));
 
 #define CREATE_ASYNC_BATON(K, VAR)                                      \
   K *VAR = new K();                                                     \
   VAR->request.data = VAR;
-
 
 /*
     create
@@ -99,7 +99,7 @@ public:
             else console.log("Error:", error);
         });
 */
-Handle<Value> create(const Arguments &args);
+NAN_METHOD(create);
 
 
 
@@ -121,7 +121,7 @@ Handle<Value> create(const Arguments &args);
             else console.log("Error:", error);
         });
 */
-Handle<Value> update(const Arguments &args);
+NAN_METHOD(update);
 
 
 
@@ -140,7 +140,7 @@ Handle<Value> update(const Arguments &args);
             console.log(time, data);
         });
 */
-Handle<Value> fetch(const Arguments &args);
+NAN_METHOD(fetch);
 
 
 
@@ -158,7 +158,7 @@ Handle<Value> fetch(const Arguments &args);
             console.log('Last updated on', new Date(time * 1000));
         });
 */
-Handle<Value> last(const Arguments &args);
+NAN_METHOD(last);
 
 
 
@@ -176,7 +176,7 @@ Handle<Value> last(const Arguments &args);
             console.log(info);
         });
 */
-Handle<Value> info(const Arguments &args); 
+NAN_METHOD(info);
 
 }
 

--- a/lib/node_rrd/src/node_rrd_create.cc
+++ b/lib/node_rrd/src/node_rrd_create.cc
@@ -1,6 +1,6 @@
 /*
     RRDtool (http://oss.oetiker.ch/rrdtool/) bindings module for node (http://nodejs.org)
-    
+
     Copyright (c), 2012 Thierry Passeron
 
     The MIT License
@@ -50,37 +50,37 @@ Infos::~Infos() {
 static void async_worker(uv_work_t *req);
 static void async_after(uv_work_t *req);
 
-Handle<Value> create(const Arguments &args) { /* rrd.create(String filename, Number step, Number start, Array spec, Function callback); */
-    HandleScope scope;
+NAN_METHOD(create) { /* rrd.create(String filename, Number step, Number start, Array spec, Function callback); */
+    Nan::HandleScope scope;
 
     CHECK_FUN_ARG(4)
 
     // Create info baton
-    CREATE_ASYNC_BATON(Infos, info)
+    CREATE_ASYNC_BATON(Infos, _info)
 
     // Get filename
-    SET_CHARS_ARG(0, info->filename)
+    SET_CHARS_ARG(0, _info->filename)
 
     // Get step
-    info->step = args[1]->Uint32Value();
+    _info->step = Nan::To<uint32_t>(info[1]).FromJust();
 
     // Get start time
-    info->time = args[2]->Uint32Value();
+    _info->time = Nan::To<uint32_t>(info[2]).FromJust();
 
     // Get spec array
-    SET_ARGC_ARGV_ARG(3, info->argc, info->argv)
+    SET_ARGC_ARGV_ARG(3, _info->argc, _info->argv)
 
     // Get callback
-    SET_PERSFUN_ARG(4, info->callback)
+    SET_PERSFUN_ARG(4, _info->callback)
 
-    uv_queue_work(uv_default_loop(), &info->request, async_worker, (uv_after_work_cb)async_after);
+    uv_queue_work(uv_default_loop(), &_info->request, async_worker, (uv_after_work_cb)async_after);
 
-    return Undefined();
+    return;
 }
 
 static void async_worker(uv_work_t *req) {
-    Infos * info = static_cast<Infos*>(req->data);
-    
+    auto info = static_cast<Infos*>(req->data);
+
     info->status = rrd_create_r(
         (const char*)info->filename,
         info->step,
@@ -91,15 +91,16 @@ static void async_worker(uv_work_t *req) {
 }
 
 static void async_after(uv_work_t *req) {
-    HandleScope scope;
+    Nan::HandleScope scope;
 
-    Infos * info = static_cast<Infos*>(req->data);
-    
-    Handle<Value> res[] = { info->status < 0 ? String::New(rrd_get_error()) : Null() };
-    info->callback->Call(Context::GetCurrent()->Global(), 1, res);
+    auto info = static_cast<Infos*>(req->data);
+
+    Local<Value> res[] = { Nan::Null() };
+    if (info->status < 0) res[0] = Nan::New<String>(rrd_get_error()).ToLocalChecked();
+    info->callback->Call(1, res);
 
     rrd_clear_error();
-    
+
     delete(info);
 }
 

--- a/lib/node_rrd/src/node_rrd_fetch.cc
+++ b/lib/node_rrd/src/node_rrd_fetch.cc
@@ -1,6 +1,6 @@
 /*
     RRDtool (http://oss.oetiker.ch/rrdtool/) bindings module for node (http://nodejs.org)
-    
+
     Copyright (c), 2012 Thierry Passeron
 
     The MIT License
@@ -36,11 +36,11 @@ public:
     time_t start;
     time_t end;        /* which time frame do you want ?
                          * will be changed to represent reality */
-    unsigned long step;    /* which stepsize do you want? 
+    unsigned long step;    /* which stepsize do you want?
                              * will be changed to represent reality */
     unsigned long ds_cnt;  /* number of data sources in file */
     char **ds_namv;    /* names of data_sources */
-    rrd_value_t *data; 
+    rrd_value_t *data;
 
     ~Infos();
 };
@@ -57,34 +57,34 @@ Infos::~Infos() {
 static void async_worker(uv_work_t *req);
 static void async_after(uv_work_t *req);
 
-Handle<Value> fetch(const Arguments &args) { // rrd.fetch(String filename, String cf, Number start, Number end, Number step, Function callback);
-    HandleScope scope;
+NAN_METHOD(fetch) { // rrd.fetch(String filename, String cf, Number start, Number end, Number step, Function callback);
+    Nan::HandleScope scope;
 
     CHECK_FUN_ARG(5)
 
     // Create baton
-    CREATE_ASYNC_BATON(Infos, info)
+    CREATE_ASYNC_BATON(Infos, _info)
 
     // Get filename
-    SET_CHARS_ARG(0, info->filename)
+    SET_CHARS_ARG(0, _info->filename)
 
     // Get template string
-    SET_CHARS_ARG(1, info->cf);
+    SET_CHARS_ARG(1, _info->cf);
 
-    info->start = args[2]->Uint32Value();
-    info->end = args[3]->Uint32Value();
-    info->step = args[4]->Uint32Value();
+    _info->start = Nan::To<uint32_t>(info[2]).FromJust();
+    _info->end = Nan::To<uint32_t>(info[3]).FromJust();
+    _info->step = Nan::To<uint32_t>(info[4]).FromJust();
 
     // Get callback
-    SET_PERSFUN_ARG(5, info->callback)
+    SET_PERSFUN_ARG(5, _info->callback)
 
-    uv_queue_work(uv_default_loop(), &info->request, async_worker, (uv_after_work_cb)async_after);
+    uv_queue_work(uv_default_loop(), &_info->request, async_worker, (uv_after_work_cb)async_after);
 
-    return Undefined();
+    return;
 }
 
 static void async_worker(uv_work_t *req) {
-    Infos * info = static_cast<Infos*>(req->data);
+    auto info = static_cast<Infos*>(req->data);
 
     info->status = rrd_fetch_r(
         (const char *)info->filename,
@@ -98,31 +98,31 @@ static void async_worker(uv_work_t *req) {
     );
 }
 
-Handle<Object> current_data_to_object(unsigned long ds_cnt, char ** ds_namv, rrd_value_t *data);
+Local<Object> current_data_to_object(unsigned long ds_cnt, char ** ds_namv, rrd_value_t *data);
 
 static void async_after(uv_work_t *req) {
-    HandleScope scope;
+    Nan::HandleScope scope;
 
-    Infos * info = static_cast<Infos*>(req->data);
-    
+    auto info = static_cast<Infos*>(req->data);
+
     if (info->status == 0) {
         rrd_value_t *datai;
         long ti;
 
         datai = info->data;
         for (ti = info->start + info->step; ti <= info->end; ti += info->step) {
-            Handle<Value> argv[] = { Number::New(ti), current_data_to_object(info->ds_cnt, info->ds_namv, datai) };
-            info->callback->Call(Context::GetCurrent()->Global(), 2, argv);
+            Local<Value> argv[] = { Nan::New<Number>(ti), current_data_to_object(info->ds_cnt, info->ds_namv, datai) };
+            info->callback->Call(2, argv);
             datai += info->ds_cnt;
         }
-        
+
         /* Last callback with (null, null) */
-        Handle<Value> argv[] = { Null(), Null() };
-        info->callback->Call(Context::GetCurrent()->Global(), 2, argv);
-        
+        Local<Value> argv[] = { Nan::Null(), Nan::Null() };
+        info->callback->Call(2, argv);
+
     } else {
-        Handle<Value> res[] = { Number::New(info->status) };
-        info->callback->Call(Context::GetCurrent()->Global(), 1, res);
+        Local<Value> res[] = { Nan::New<Number>(info->status) };
+        info->callback->Call(1, res);
     }
 
     delete(info);
@@ -135,21 +135,21 @@ static void async_after(uv_work_t *req) {
         ...
     }
 */
-Handle<Object> current_data_to_object(unsigned long ds_cnt, char ** ds_namv, rrd_value_t *data) { 
-    HandleScope scope;
+Local<Object> current_data_to_object(unsigned long ds_cnt, char ** ds_namv, rrd_value_t *data) {
+    Nan::EscapableHandleScope scope;
 
-    Handle<ObjectTemplate> obj = ObjectTemplate::New();
-    Handle<Object> result = obj->NewInstance();
+    auto obj = ObjectTemplate::New();
+    auto result = obj->NewInstance();
 
     unsigned long ii;
     rrd_value_t *datai;
 
     datai = data;
     for (ii = 0; ii < ds_cnt; ii++) {
-        result->Set(String::New(ds_namv[ii]), Number::New(datai[ii]));
+        result->Set(Nan::New<String>(ds_namv[ii]).ToLocalChecked(), Nan::New<Number>(datai[ii]));
     }
 
-    return scope.Close(result);
+    return scope.Escape(result);
 }
 
 }

--- a/lib/node_rrd/src/node_rrd_info.cc
+++ b/lib/node_rrd/src/node_rrd_info.cc
@@ -1,6 +1,6 @@
 /*
     RRDtool (http://oss.oetiker.ch/rrdtool/) bindings module for node (http://nodejs.org)
-    
+
     Copyright (c), 2012 Thierry Passeron
 
     The MIT License
@@ -34,7 +34,7 @@ class Infos: public AsyncInfos {
 public:
     rrd_info_t *data;
 
-    ~Infos() { free(data); }
+    ~Infos() { rrd_info_free(data); }
 };
 
 }
@@ -42,92 +42,75 @@ public:
 static void async_worker(uv_work_t *req);
 static void async_after(uv_work_t *req);
 
-Handle<Value> info(const Arguments &args) { // rrd.info(String filename, Function callback);
-    HandleScope scope;
+NAN_METHOD(info) { // rrd.info(String filename, Function callback);
+    Nan::HandleScope scope;
 
     CHECK_FUN_ARG(1)
 
     // Create info baton
-    CREATE_ASYNC_BATON(Infos, info)
+    CREATE_ASYNC_BATON(Infos, _info)
 
     // Get filename
-    SET_CHARS_ARG(0, info->filename)
+    SET_CHARS_ARG(0, _info->filename)
 
     // Get callback
-    SET_PERSFUN_ARG(1, info->callback)
+    SET_PERSFUN_ARG(1, _info->callback)
 
-    uv_queue_work(uv_default_loop(), &info->request, async_worker, (uv_after_work_cb)async_after);
+    uv_queue_work(uv_default_loop(), &_info->request, async_worker, (uv_after_work_cb)async_after);
 
-    return Undefined();
+    return;
 }
 
 static void async_worker(uv_work_t *req) {
-    Infos * info = static_cast<Infos*>(req->data);
+    auto info = static_cast<Infos*>(req->data);
 
     info->data = rrd_info_r(info->filename);
 }
 
-static Handle<Object>info_data_to_object(rrd_info_t *data);
+static Local<Object>info_data_to_object(rrd_info_t *data);
 
 static void async_after(uv_work_t *req) {
-    HandleScope scope;
+    Nan::HandleScope scope;
 
-    Infos * info = static_cast<Infos*>(req->data);
-    
-    Handle<Value> res[] = { info_data_to_object((rrd_info_t *)info->data) };
-    info->callback->Call(Context::GetCurrent()->Global(), 1, res);
+    auto info = static_cast<Infos*>(req->data);
 
-    delete(info);    
+    Local<Value> res[] = { info_data_to_object((rrd_info_t *)info->data) };
+    info->callback->Call(1, res);
+
+    delete(info);
 }
 
-/*
- * Based off of http://sambro.is-super-awesome.com/2011/03/03/creating-a-proper-buffer-in-a-node-c-addon/
- */
-static Local<Object> makeBuffer(char* data, size_t size) {
-  HandleScope scope;
+static Local<Object>info_data_to_object(rrd_info_t *data) {
+    Nan::EscapableHandleScope scope;
 
-  Local<Buffer> slowBuffer = Buffer::New(data, size);
-  Local<Object> global = Context::GetCurrent()->Global();
-  Local<Value> bv = global->Get(String::NewSymbol("Buffer"));
-  assert(bv->IsFunction());
-  Local<Function> b = Local<Function>::Cast(bv);
-  Handle<Value> argv[3] = { slowBuffer->handle_, Integer::New(size), Integer::New(0) };
-  Local<Object> fastBuffer = b->NewInstance(3, argv);
+    auto info = ObjectTemplate::New();
+    auto instance = info->NewInstance();
 
-  return scope.Close(fastBuffer);
-}
-
-static Handle<Object>info_data_to_object(rrd_info_t *data) {
-    HandleScope scope;
-
-    Handle<ObjectTemplate> info = ObjectTemplate::New();
-    Handle<Object> instance = info->NewInstance();
-    
     while (data) {
-         v8::Handle<v8::String> key = String::New(data->key);
+         v8::Local<v8::String> key = Nan::New<String>(data->key).ToLocalChecked();
 
         switch (data->type) {
         case RD_I_VAL:
-            instance->Set(key, Number::New(data->value.u_val)); // if data->value.u_val is not a number, it will be NaN
+            instance->Set(key, Nan::New<Number>(data->value.u_val)); // if data->value.u_val is not a number, it will be NaN
             break;
         case RD_I_CNT:
-            instance->Set(key, Number::New(data->value.u_cnt));
+            instance->Set(key, Nan::New<Number>(data->value.u_cnt));
             break;
         case RD_I_INT:
-            instance->Set(key, Number::New(data->value.u_int));
+            instance->Set(key, Nan::New<Number>(data->value.u_int));
             break;
         case RD_I_STR:
-            instance->Set(key, String::New(data->value.u_str));
+            instance->Set(key, Nan::New<String>(data->value.u_str).ToLocalChecked());
             break;
         case RD_I_BLO:
-            instance->Set(key, makeBuffer((char*)data->value.u_blo.ptr, data->value.u_blo.size));
+            instance->Set(key, Nan::NewBuffer((char*)data->value.u_blo.ptr, data->value.u_blo.size).ToLocalChecked());
             // fwrite(data->value.u_blo.ptr, data->value.u_blo.size, 1, stdout);
             break;
         }
         data = data->next;
     }
 
-    return scope.Close(instance);
+    return scope.Escape(instance);
 }
 
 }

--- a/lib/node_rrd/src/node_rrd_last.cc
+++ b/lib/node_rrd/src/node_rrd_last.cc
@@ -1,6 +1,6 @@
 /*
     RRDtool (http://oss.oetiker.ch/rrdtool/) bindings module for node (http://nodejs.org)
-    
+
     Copyright (c), 2012 Thierry Passeron
 
     The MIT License
@@ -40,39 +40,39 @@ public:
 static void async_worker(uv_work_t *req);
 static void async_after(uv_work_t *req);
 
-Handle<Value> last(const Arguments &args) { // rrd.last(String filename, Function callback);
-    HandleScope scope;
+NAN_METHOD(last) { // rrd.last(String filename, Function callback);
+    Nan::HandleScope scope;
 
     CHECK_FUN_ARG(1)
 
     // Create info baton
-    CREATE_ASYNC_BATON(Infos, info)
+    CREATE_ASYNC_BATON(Infos, _info)
 
     // Get filename
-    SET_CHARS_ARG(0, info->filename)
+    SET_CHARS_ARG(0, _info->filename)
 
     // Get callback
-    SET_PERSFUN_ARG(1, info->callback)
+    SET_PERSFUN_ARG(1, _info->callback)
 
-    uv_queue_work(uv_default_loop(), &info->request, async_worker, (uv_after_work_cb)async_after);
+    uv_queue_work(uv_default_loop(), &_info->request, async_worker, (uv_after_work_cb)async_after);
 
-    return Undefined();
+    return;
 }
 
 static void async_worker(uv_work_t *req) {
-    Infos * info = static_cast<Infos*>(req->data);
+    auto info = static_cast<Infos*>(req->data);
 
     info->last = rrd_last_r(info->filename);
 }
 
 static void async_after(uv_work_t *req) {
-    HandleScope scope;
+    Nan::HandleScope scope;
 
-    Infos * info = static_cast<Infos*>(req->data);
-    
-    Handle<Value> res[] = { Number::New(info->last) };
-    info->callback->Call(Context::GetCurrent()->Global(), 1, res);
-    
+    auto info = static_cast<Infos*>(req->data);
+
+    Local<Value> res[] = { Nan::New<Number>(info->last) };
+    info->callback->Call(1, res);
+
     delete(info);
 }
 

--- a/lib/node_rrd/src/node_rrd_update.cc
+++ b/lib/node_rrd/src/node_rrd_update.cc
@@ -1,6 +1,6 @@
 /*
     RRDtool (http://oss.oetiker.ch/rrdtool/) bindings module for node (http://nodejs.org)
-    
+
     Copyright (c), 2012 Thierry Passeron
 
     The MIT License
@@ -50,33 +50,33 @@ Infos::~Infos() {
 static void async_worker(uv_work_t *req);
 static void async_after(uv_work_t *req);
 
-Handle<Value> update(const Arguments &args) { // rrd.update(String filename, String tmplt, Array updates, Function callback);
-    HandleScope scope;
+NAN_METHOD(update) { // rrd.update(String filename, String tmplt, Array updates, Function callback);
+    Nan::HandleScope scope;
 
     CHECK_FUN_ARG(3)
 
     // Create info baton
-    CREATE_ASYNC_BATON(Infos, info)
+    CREATE_ASYNC_BATON(Infos, _info)
 
     // Get filename
-    SET_CHARS_ARG(0, info->filename)
+    SET_CHARS_ARG(0, _info->filename)
 
     // Get template string
-    SET_CHARS_ARG(1, info->tmplt);
+    SET_CHARS_ARG(1, _info->tmplt);
 
     // Get updates array
-    SET_ARGC_ARGV_ARG(2, info->argc, info->argv)
+    SET_ARGC_ARGV_ARG(2, _info->argc, _info->argv)
 
     // Get callback
-    SET_PERSFUN_ARG(3, info->callback)
+    SET_PERSFUN_ARG(3, _info->callback)
 
-    uv_queue_work(uv_default_loop(), &info->request, async_worker, (uv_after_work_cb)async_after);
+    uv_queue_work(uv_default_loop(), &_info->request, async_worker, (uv_after_work_cb)async_after);
 
-    return Undefined();
+    return;
 }
 
 static void async_worker(uv_work_t *req) {
-    Infos * info = static_cast<Infos*>(req->data);
+    auto info = static_cast<Infos*>(req->data);
 
     int status = rrd_update_r(
         (const char *)info->filename,
@@ -89,15 +89,16 @@ static void async_worker(uv_work_t *req) {
 }
 
 static void async_after(uv_work_t *req) {
-    HandleScope scope;
+    Nan::HandleScope scope;
 
-    Infos * info = static_cast<Infos*>(req->data);
-    
-    Handle<Value> res[] = { info->status < 0 ? String::New(rrd_get_error()) : Null() };
-    info->callback->Call(Context::GetCurrent()->Global(), 1, res);
+    auto info = static_cast<Infos*>(req->data);
+
+    Local<Value> res[] = { Nan::Null() };
+    if (info->status < 0) res[0] = Nan::New<String>(rrd_get_error()).ToLocalChecked();
+    info->callback->Call(1, res);
 
     rrd_clear_error();
-    
+
     delete(info);
 }
 

--- a/lib/rrd/rrddir.js
+++ b/lib/rrd/rrddir.js
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 var walk = require('walk');
-var rrd = require('../node_rrd/lib/rrd.js');
+var rrd = require('rrd');
 
 exports.getKeys = getKeys;
 

--- a/setup.sh
+++ b/setup.sh
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -e
+
 . ./env.sh
 
 export LD_LIBRARY_PATH=$COLLECTD_HOME/lib
@@ -24,7 +26,6 @@ CONFIG_FILES=(
   conf/conf-orbiter.json
   conf/conf-rrd.json
   conf/conf-view.json
-  lib/node_rrd/binding.gyp
   view/modules/config.js
 )
 
@@ -34,17 +35,13 @@ for file in ${CONFIG_FILES[@]}; do
   TEMPLATE $file
 done
 
-# Build node_rrd
-pushd lib/node_rrd
-rm -rf build
-rm -rf node_modules
-npm install .
-popd
-
 # Build hubble
 rm -rf node_modules
 npm install -g forever
 npm install
+
+# Install node_rrd
+npm install $(pwd)/lib/node_rrd
 
 # Copy collectd stuffs
 mkdir -p $COLLECTD_HOME/plugins/python

